### PR TITLE
Updating Packages to Match

### PR DESCRIPTION
**Pull request by mfixstsci**

### CODE REVIEW

**Comment by mfixstsci** [comment]
| Package          | jwql-3.8 | jwql-3.9 | Requirements | Match |
| ---------------- | -------- | -------- | ------------ | ----- |
| asdf             |  2.11.0  |  2.8.3   |    2.11.0    | false |
| astropy          |  5.0.2   |  4.3.1   |    5.0.2     | false |
| astroquery       |  0.4.5   |  0.4.5   |    0.4.5     | true  |
| bandit           |  1.7.4   |  1.7.1   |    1.7.4     | false |
| bokeh            |  2.4.2   |  2.4.2   |    2.4.2     | true  |
| beautifulsoup4   |  4.10.0  |  4.9.3   |    4.10.0    | false |
| codecov          |  2.1.11  |  2.1.11  |    2.1.12    | false |
| cryptography     |  36.0.0  |  36.0.0  |    36.0.0    | true  |
| django           |  3.2.5   |  3.1.7   |    3.2.5     | false |
| flake8           |  3.9.2   |  3.9.2   |    4.0.1     | false |
| inflection       |  0.5.1   |  0.5.1   |    0.5.1     | true  |
| ipython          |  7.29.0  |  7.29.0  |    8.1.0     | false |
| jinja2           |  3.0.2   |  3.0.2   |    3.0.3     | false |
| jsonschema       |  3.2.0   |  3.2.0   |    4.4.0     | false |
| jwst             |  1.4.6   |  1.4.6   |    1.4.6     | true  |
| matplotlib       |  3.4.2   |  3.4.2   |    3.5.1     | false |
| nodejs           |  10.13.0 |  10.13.0 |    10.13.0   | true  |
| numpy            |  1.20.3  |  1.20.3  |    1.22.2    | false |
| numpydoc         |  1.1.0   |  1.1.0   |    1.2       | false |
| pandas           |  1.3.4   |  1.3.4   |    1.4.1     | false |
| pip              |  21.0.1  |  21.0.1  |    N/A       | true  |
| postgresql       |  12.2    |  12.2    |    N/A       | true  |
| psycopg2         |  2.8.6   |  2.8.6   |    2.9.3     | false |
| pysiaf           |  0.15.0  |  0.15.0  |    0.15.0    | true  |
| pysqlite3        |  0.4.3   |  0.4.3   |    N/A       | true  |
| pytest           |  6.2.4   |  6.2.4   |    7.0.1     | false |
| pytest-cov       |  3.0.0   |  3.0.0   |    3.0.0     | true  |
| python           |  3.8.5   |  3.9.11  |    N/A       | false |
| pyyaml           |  5.4.1   |  5.4.1   |    N/A       | true  |
| scipy            |  1.7.1   |  1.7.1   |    1.8.0     | false |
| setuptools       |  52.0.0  |  52.0.0  |    N/A       | true  |
| sphinx           |  4.2.0   |  4.2.0   |    4.4.0     | false |
| sphinx_rtd_theme |  0.4.3   |  0.4.3   |    1.0.0     | false |
| sqlalchemy       |  1.4.27  |  1.4.27  |    1.4.31    | false |
| stsci_rtd_theme  |  0.0.2   |  0.0.2   |    0.0.2     | true  |
| twine            |  3.4.1   |  3.4.1   |    3.8.0     | false |
| wtforms          |  2.3.3   |  2.3.3   |    3.0.1     | false |

**Comment by mfixstsci** [comment]
| Package          | jwql-3.8 | jwql-3.9 | Requirements | Match |
| ---------------- | -------- | -------- | ------------ | ----- |
| astropy          |  5.0.4   |  5.0.4   |    5.0.4     | true  |
| astroquery       |  0.4.6   |  0.4.6   |    0.4.6     | true  |
| bandit           |  1.7.4   |  1.7.4   |    1.7.4     | true  |
| bokeh            |  2.4.2   |  2.4.2   |    2.4.2     | true  |
| beautifulsoup4   |  4.9.3   |  4.9.3   |    4.9.3     | true  |
| codecov          |  2.1.11  |  2.1.11  |    2.1.11    | true  |
| cryptography     |  36.0.0  |  36.0.0  |    36.0.0    | true  |
| django           |  3.2.5   |  3.2.5   |    3.2.5     | true  |
| flake8           |  4.0.1   |  4.0.1   |    4.0.1     | true  |
| inflection       |  0.5.1   |  0.5.1   |    0.5.1     | true  |
| ipython          |  8.2.0   |  8.2.0   |    8.2.0     | true  |
| jinja2           |  3.0.3   |  3.0.3   |    3.0.3     | true  |
| jsonschema       |  4.4.0   |  4.4.0   |    4.4.0     | true  |
| jwst             |  1.4.6   |  1.4.6   |    1.4.6     | true  |
| matplotlib       |  3.4.2   |  3.4.2   |    3.4.2     | true  |
| nodejs           |  10.13.0 |  10.13.0 |    10.13.0   | true  |
| numpy            |  1.20.3  |  1.20.3  |    1.20.3    | true  |
| numpydoc         |  1.2     |  1.2     |    1.2       | true  |
| pandas           |  1.4.2   |  1.4.2   |    1.4.2     | true  |
| pip              |  21.2.4  |  21.2.4  |    N/A       | true  |
| postgresql       |  12.2    |  12.2    |    N/A       | true  |
| psycopg2         |  2.8.6   |  2.8.6   |    2.8.6     | true  |
| pysiaf           |  0.15.0  |  0.15.0  |    0.15.0    | true  |
| pysqlite3        |  0.4.7   |  0.4.7   |    N/A       | true  |
| pytest           |  7.1.1   |  7.1.1   |    7.1.1     | true  |
| pytest-cov       |  3.0.0   |  3.0.0   |    3.0.0     | true  |
| python           |  3.8.5   |  3.9.11  |    N/A       | N/A   |
| pyyaml           |  6.0     |  6.0     |    N/A       | true  |
| scipy            |  1.7.3   |  1.7.3   |    1.7.3     | true  |
| setuptools       |  52.0.0  |  52.0.0  |    N/A       | true  |
| sphinx           |  4.4.0   |  4.4.0   |    4.4.0     | true  |
| sphinx_rtd_theme |  0.4.3   |  0.4.3   |    0.4.3     | true  |
| sqlalchemy       |  1.4.32  |  1.4.32  |    1.4.32    | true  |
| stsci_rtd_theme  |  0.0.2   |  0.0.2   |    0.0.2     | true  |
| twine            |  3.7.1   |  3.7.1   |    3.7.1     | true  |
| wtforms          |  2.3.3   |  2.3.3   |    2.3.3     | true  |

All of the package numbers should match in the files all of our `env` and `requirements.txt` files.

**Comment by mfixstsci** [comment]
| Package          | jwql-3.8 | jwql-3.9 | Requirements | Match |
| ---------------- | -------- | -------- | ------------ | ----- |
| astropy          |  5.0.4   |  5.0.4   |    5.0.4     | true  |
| astroquery       |  0.4.6   |  0.4.6   |    0.4.6     | true  |
| bandit           |  1.7.4   |  1.7.4   |    1.7.4     | true  |
| bokeh            |  2.4.2   |  2.4.2   |    2.4.2     | true  |
| beautifulsoup4   |  4.9.3   |  4.9.3   |    4.9.3     | true  |
| codecov          |  2.1.11  |  2.1.11  |    2.1.11    | true  |
| cryptography     |  36.0.0  |  36.0.0  |    36.0.0    | true  |
| django           |  3.2.5   |  3.2.5   |    3.2.5     | true  |
| flake8           |  4.0.1   |  4.0.1   |    4.0.1     | true  |
| inflection       |  0.5.1   |  0.5.1   |    0.5.1     | true  |
| ipython          |  8.2.0   |  8.2.0   |    8.2.0     | true  |
| jinja2           |  3.0.3   |  3.0.3   |    3.0.3     | true  |
| jsonschema       |  4.4.0   |  4.4.0   |    4.4.0     | true  |
| jwst             |  1.4.6   |  1.4.6   |    1.4.6     | true  |
| matplotlib       |  3.4.2   |  3.4.2   |    3.4.2     | true  |
| nodejs           |  0.1.1   |  0.1.1   |    0.1.1     | true  |
| numpy            |  1.22.3  |  1.22.3  |    1.22.3    | true  |
| numpydoc         |  1.2     |  1.2     |    1.2       | true  |
| pandas           |  1.4.2   |  1.4.2   |    1.4.2     | true  |
| pip              |  21.2.4  |  21.2.4  |    N/A       | true  |
| postgresql       |  12.2    |  12.2    |    N/A       | true  |
| psycopg2         |  2.8.6   |  2.8.6   |    2.8.6     | true  |
| pysiaf           |  0.15.0  |  0.15.0  |    0.15.0    | true  |
| pysqlite3        |  0.4.7   |  0.4.7   |    N/A       | true  |
| pytest           |  7.1.1   |  7.1.1   |    7.1.1     | true  |
| pytest-cov       |  3.0.0   |  3.0.0   |    3.0.0     | true  |
| python           |  3.8.5   |  3.9.11  |    N/A       | N/A   |
| pyyaml           |  6.0     |  6.0     |    N/A       | true  |
| scipy            |  1.7.3   |  1.7.3   |    1.7.3     | true  |
| setuptools       |  52.0.0  |  52.0.0  |    N/A       | true  |
| sphinx           |  4.4.0   |  4.4.0   |    4.4.0     | true  |
| sphinx_rtd_theme |  0.4.3   |  0.4.3   |    0.4.3     | true  |
| sqlalchemy       |  1.4.32  |  1.4.32  |    1.4.32    | true  |
| stsci_rtd_theme  |  0.0.2   |  0.0.2   |    0.0.2     | true  |
| twine            |  3.7.1   |  3.7.1   |    3.7.1     | true  |
| wtforms          |  2.3.3   |  2.3.3   |    2.3.3     | true  |

**Comment by bhilbert4** [comment]
The big change in nodejs had me suspicious. Looks like 0.1.1 is where they gave up on hosting on pypi. But using conda, the latest version in the `pkgs/main` channel is 16.13.1. I haven't yet found a list of dates for each version, but it looks like maybe the 10.13 version we had been using is fairly old?

**Comment by mfixstsci** [comment]
| Package          | jwql-3.8 | jwql-3.9 | Requirements | Match |
| ---------------- | -------- | -------- | ------------ | ----- |
| astropy          |  5.0.4   |  5.0.4   |    5.0.4     | true  |
| astroquery       |  0.4.6   |  0.4.6   |    0.4.6     | true  |
| bandit           |  1.7.4   |  1.7.4   |    1.7.4     | true  |
| bokeh            |  2.4.2   |  2.4.2   |    2.4.2     | true  |
| beautifulsoup4   |  4.9.3   |  4.9.3   |    4.9.3     | true  |
| codecov          |  2.1.11  |  2.1.11  |    2.1.11    | true  |
| cryptography     |  36.0.0  |  36.0.0  |    36.0.0    | true  |
| django           |  3.2.5   |  3.2.5   |    3.2.5     | true  |
| flake8           |  4.0.1   |  4.0.1   |    4.0.1     | true  |
| inflection       |  0.5.1   |  0.5.1   |    0.5.1     | true  |
| ipython          |  8.2.0   |  8.2.0   |    8.2.0     | true  |
| jinja2           |  3.0.3   |  3.0.3   |    3.0.3     | true  |
| jsonschema       |  4.4.0   |  4.4.0   |    4.4.0     | true  |
| jwst             |  1.4.6   |  1.4.6   |    1.4.6     | true  |
| matplotlib       |  3.4.2   |  3.4.2   |    3.4.2     | true  |
| nodejs           |  10.13.0   |  10.13.0  |    0.1.1     | False  |
| numpy            |  1.21.5  |  1.21.5  |    1.21.5    | true  |
| numpydoc         |  1.2     |  1.2     |    1.2       | true  |
| pandas           |  1.4.2   |  1.4.2   |    1.4.2     | true  |
| pip              |  21.2.4  |  21.2.4  |    N/A       | true  |
| postgresql       |  12.2    |  12.2    |    N/A       | true  |
| psycopg2         |  2.8.6   |  2.8.6   |    2.8.6     | true  |
| pysiaf           |  0.15.0  |  0.15.0  |    0.15.0    | true  |
| pysqlite3        |  0.4.7   |  0.4.7   |    N/A       | true  |
| pytest           |  7.1.1   |  7.1.1   |    7.1.1     | true  |
| pytest-cov       |  3.0.0   |  3.0.0   |    3.0.0     | true  |
| python           |  3.8.5   |  3.9.11  |    N/A       | N/A   |
| pyyaml           |  6.0     |  6.0     |    N/A       | true  |
| scipy            |  1.7.3   |  1.7.3   |    1.7.3     | true  |
| setuptools       |  52.0.0  |  52.0.0  |    N/A       | true  |
| sphinx           |  4.4.0   |  4.4.0   |    4.4.0     | true  |
| sphinx_rtd_theme |  0.4.3   |  0.4.3   |    0.4.3     | true  |
| sqlalchemy       |  1.4.32  |  1.4.32  |    1.4.32    | true  |
| stsci_rtd_theme  |  0.0.2   |  0.0.2   |    0.0.2     | true  |
| twine            |  3.7.1   |  3.7.1   |    3.7.1     | true  |
| wtforms          |  2.3.3   |  2.3.3   |    2.3.3     | true  |

**Comment by mfixstsci** [comment]
Looks like the tests are passing here now too 🎉 just need a review. I downgraded some files in the pip to match our conda env to try and match 1-to-1. I can reread the wiki page on this and if keeping the versions to the latest release is the practice we want to continue then I can revert them back.

**Comment by bhilbert4** [comment]
I guess the question is, who, if anyone, is using requirements.txt? Github Actions is using the env files right? And RTD has its own requirements file. The servers and team members are using the env files. If it turns out no one is really using requirements.txt, then it probably doesn't make sense to hold back versions of certain dependencies in the env files just to have them match requirements.txt, right? What really gives me pause is nodejs. If someone built an environment and used version 0.1.1, would jwql even work?

**Comment by mfixstsci** [comment]
@bhilbert4 I agree with this, it might also be a hold over from when we used Travis CI. I believe that Travis liked using `requirements.txt` files and then we moved over to conda envs when we started using github actions CI. I can dig up some old versions of the CI to verify.

**Comment by mfixstsci** [comment]
Looks like we used `JENKINS` CI early on.

https://github.com/spacetelescope/jwql/blob/v0.10.4/Jenkinsfile#L8

But! It was still using a conda env file.

**Comment by mfixstsci** [comment]
Looks like the `requirements.txt` file first appears in v0.16.0 @bhilbert4 

https://github.com/spacetelescope/jwql/blob/v0.16.0/requirements.txt

I think this might a "I couldn't see the forest through the trees" moment. I think that the dependency updating bots we have hooked into the repo check the requirements file to keep us up-to-date on our dependencies. See the first commit below:

https://github.com/spacetelescope/jwql/commit/bb972ef93ba345800040e3c434bbb63860e44823

**Comment by mfixstsci** [comment]
If we look at the history from this point: 

https://github.com/spacetelescope/jwql/commits/v0.16.0/requirements.txt

@bourque made a few commits to get the file up and running with `pyup` and then the bot made most of the changes after that.

**Comment by mfixstsci** [comment]
Hey @bhilbert4 and @BradleySappington what do we think of this? I would like to get this in soon :)

**Comment by BradleySappington** [comment]
Seems like thats as close as we can get right? And Requirements.txt isn't really going to be used.  Seems okay to me but I'd defer to @bhilbert4 as owner

**Comment by mfixstsci** [comment]
@BradleySappington it is used by `pyup/dependabot` to check the most up-to-date versions of our dependencies via pip. Since we use the conda `.env` to install our dependencies, we need the requiements file to get checked by the dependency hooks we have set and then let that guide us in updating our conda `.env` files. So they don't have to match one-to-one but serve more as a guide.

The main reason for this PR is because there were mismatching versions of our software in the `3.8` and `3.9` files that should have been matching or close to.

I am hoping one day there is a conda based solution, then we could delete the requirements file once and for all 🎉

**Comment by bourque** [comment]
> I am hoping one day there is a conda based solution, then we could delete the requirements file once and for all 🎉

https://github.com/pyupio/pyup/issues/322

Please like and subscribe 🔥 💯

**Comment by mfixstsci** [comment]
@bourque I AM SCREAMING 🤣

**Comment by bhilbert4** [comment]
Last comment there was from Sept 2020. Where's the holding-my-breath emoji?

So pyup pays attention to the packages in the pip section of the environment files and for everything else, it looks in `requirements.txt`? Is that right? In that case, would it make sense for us to move as many packages as possible into the pip section of our env files? Would that help minimize potential inconsistencies?

**Comment by mfixstsci** [comment]
> Last comment there was from Sept 2020. Where's the holding-my-breath emoji?
> 
> So pyup pays attention to the packages in the pip section of the environment files and for everything else, it looks in `requirements.txt`? Is that right? In that case, would it make sense for us to move as many packages as possible into the pip section of our env files? Would that help minimize potential inconsistencies?

Hey -- I like your thinking but fundamentally that beats the purpose of conda as a dependency and package management system because they look and pull dependencies from different places. We only use the pip section as a last resort because the software isn't available in conda forge or whatever channels we have set.

You can see which conda channels you are pulling from in you `.condarc` file in your home directory:

`$ open ~/.condarc`

If you want a deep dive: https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html

**Comment by bhilbert4** [comment]
Oops. Is my pip bias showing? 😄 
conda seems to get confused more often when it comes to dependencies and conflicts, which is why I generally prefer pip. 

Anyway, if this is the best we can do here at the moment, then so be it.

**Comment by bhilbert4** [comment]
I also remember conversations with a jwst developer a while back who warned against using conda-forge, although it looks like those conversations are now gone since the developer has left STScI.

**Comment by mfixstsci** [comment]
@bhilbert4 that's fine, mine are set to `ssb-astroconda` and then the `default`'s. What do you have listed in your `.condarc`?

File: environment_python_3_8.yml [hunk 1]
```diff
@@ -21,44 +21,43 @@ channels:
   - defaults
 
 dependencies:
-  - astropy=5.0.2
+  - astropy=5.0.4
   - bokeh=2.4.2
-  - beautifulsoup4=4.10.0
+  - beautifulsoup4=4.9.3
```
Comment (BradleySappington) [inline environment_python_3_8.yml]: Beautiful soup is moving down (and 3.8 had a higher version than 3.9?)

**Review by BradleySappington — APPROVED** [review]

File: environment_python_3_8.yml (hunk 1 above)
Comment (mfixstsci) [reply environment_python_3_8.yml]: Good point, I just looked and I could technically update it to `4.11.1` for `python 3.8` 😱 this probably means it's also available in `python 3.9` as well.

Comment (mfixstsci) [reply environment_python_3_8.yml]: Updated it in the env and requirements files 👍

**Comment by mfixstsci** [comment]
Thanks for the review @BradleySappington! Merging

**Review by BradleySappington — APPROVED** [review]
Looks great @mfixstsci

**Pull request merged**